tests/python_package_test/test_dask.py

```diff
@@ -1464,8 +1464,7 @@ def test_init_score(task, output, cluster):
             init_scores = dy.map_blocks(lambda x: np.full((x.size, size_factor), init_score))
         model = model_factory(client=client, **params)
         model.fit(dX, dy, sample_weight=dw, init_score=init_scores, group=dg)
-        # value of the root node is 0 when init_score is set
-        assert model.booster_.trees_to_dataframe()["value"][0] == 0
+        assert model.fitted_
```

Comment:
Can you help me understand why this test had to be changed, when in `gbdt.cpp` I see the following?

```cpp
// extend init_scores with zeros
        new_tree->AsConstantTree(0, num_data_);
```

And if this is not longer the case... can you please help us come up with a better way to change this test? This particular test is called `test_init_score` .... it's specific goal is to check that when you provide `init_score` to the `.fit()` method for the Dask estimators, that `init_score` actually makes it all the way down to the Booster and affects the resulting model.

Changing this condition to just "model fitting did not raise an error" is not sufficiently strict... that test would pass if `DaskLGBMRegressor.fit()` simply ignore `init_score`, for example.

Reply:
Reverted! 

Nice catch, seems like it got fixed in the meantime.

Reply:
> *it now does not execute on LGBMRanker. However this is out-of-scope for this PR*

If the changes in your PR require tests that were previously passing to now be skipped, the need to skip those tests is not "out-of-scope".

Can you please explain here what issues you saw with `LGBMRanker` and this PR's changes?

Reply:
The way the `init_score` was tested seems like a flawed way to test it, because as we see from this PR the root value was always 0. 

Rewriting the test to be "check if the 2 models are different" seems to have more merit, since if you have an init score you would expect that to happen with a small number of trees. 

It seems however that is not the case in the `LGBMRanker`, i.e. when boosting from an `init_score` or without one the models are exactly the same. Could you tell me whether I am missing something and this is the expected functionality?

Reply:
I am still not totally sure what the problem is. There doesn't seem to exist a test for normal `ranking` with `init_score`.

Maybe the problem is in the way the test is written or something related to that, but I can still not pinpoint what exactly is happening.

Reply:
After a little bit more investigation, the problem seems to be related to how ranking actually works. 
As I was not familiar with this type of task, I assumed having a constant `init_score` for all data points made sense as that was what the test was doing. This however seems not to be the case as it is subjective on the other `score`-s. Now the test should be properly testing all cases, can you confirm @jameslamb? Sorry for all the confusion caused!

Reply:
> *The way the init_score was tested seems like a flawed way to test it, because as we see from this PR the root value was always 0.*

I finally was able to test this... this claim is just not true.

I just tested on latest `master` (https://github.com/microsoft/LightGBM/commit/06432300c0c01268c8a80c3537eef81dd5ede30d), with a patch like this:

```diff
diff --git a/tests/python_package_test/test_dask.py b/tests/python_package_test/test_dask.py
index 247f2eb1..e057320b 100644
--- a/tests/python_package_test/test_dask.py
+++ b/tests/python_package_test/test_dask.py
@@ -1463,7 +1463,7 @@ def test_init_score(task, output, cluster):
         else:
             init_scores = dy.map_blocks(lambda x: np.full((x.size, size_factor), init_score))
         model = model_factory(client=client, **params)
-        model.fit(dX, dy, sample_weight=dw, init_score=init_scores, group=dg)
+        model.fit(dX, dy, sample_weight=dw, group=dg)
         # value of the root node is 0 when init_score is set
         assert model.booster_.trees_to_dataframe()["value"][0] == 0
```

<details><summary>testing code (click me)</summary>

```shell
docker run \
    --rm \
    -v $(pwd):/opt/work \
    -w /opt/work \
    -it python:3.10 \
    bash

sh build-python.sh bdist_wheel install

pip install \
    cloudpickle \
    dask \
    distributed \
    numpy \
    pandas \
    pyarrow \
    pytest \
    scikit-learn \
    scipy

pytest 'tests/python_package_test/test_dask.py::test_init_score'
```
</details>

With `init_score` not provided, the root node's `value` is a non-0 value, and the test reliably and consistently fail.

```text
E           assert np.float64(-0.0645985) == 0

tests/python_package_test/test_dask.py:1468: AssertionError
```

I will try to figure out how that got there, because now it's making me question some of the other changes in this PR.

Reply:
Alright, I *think* I found it. I think this is all happening in `GBDT::TrainOneIter()`.

That function is called once per boosting round.

As you found (and as seen in the diff of this PR), when it runs for the first boosting round, it uses `AsConstantTree()` to create a stump with the internal value set from a vector named `init_scores`.

https://github.com/microsoft/LightGBM/blob/06432300c0c01268c8a80c3537eef81dd5ede30d/src/boosting/gbdt.cpp#L430

`init_scores` is created near the beginning of the function, 

https://github.com/microsoft/LightGBM/blob/06432300c0c01268c8a80c3537eef81dd5ede30d/src/boosting/gbdt.cpp#L346

That has special logic to detect whether this is the first round of boosting...

https://github.com/microsoft/LightGBM/blob/06432300c0c01268c8a80c3537eef81dd5ede30d/src/boosting/gbdt.cpp#L347-L352

...and that logic sets the value of the root node to the "average" of the loss function evaluated over all samples if no `init_score` is provided...

https://github.com/microsoft/LightGBM/blob/06432300c0c01268c8a80c3537eef81dd5ede30d/src/boosting/gbdt.cpp#L322

... or `0.0` if an `init_score` was provided

https://github.com/microsoft/LightGBM/blob/06432300c0c01268c8a80c3537eef81dd5ede30d/src/boosting/gbdt.cpp#L341

This is why that Dask test of "root value set to 0" was a reliable way to detect that an `init_score` had been provided and made it all the way through to the `Dataset` on the C++ side.

To test my understanding, I applied a patch like this:

```diff
diff --git a/src/boosting/gbdt.cpp b/src/boosting/gbdt.cpp
index a58f34e9..94b7c4fe 100644
--- a/src/boosting/gbdt.cpp
+++ b/src/boosting/gbdt.cpp
@@ -338,7 +338,7 @@ double GBDT::BoostFromAverage(int class_id, bool update_scorer) {
       Log::Warning("Disabling boost_from_average in %s may cause the slow convergence", objective_function_->GetName());
     }
   }
-  return 0.0f;
+  return 0.56789f;
 }
 
 bool GBDT::TrainOneIter(const score_t* gradients, const score_t* hessians) {
```

Then rebuilt the library and ran this test. Saw exactly that value I hard-coded into `GBDT::BoostFromAverage()` getting set as the root node's value when `init_score` was provided.

```shell
pytest 'tests/python_package_test/test_dask.py::test_init_score'
```

```text
FAILED tests/python_package_test/test_dask.py::test_init_score[array-binary-classification] - assert np.float64(0.56789) == 0
FAILED tests/python_package_test/test_dask.py::test_init_score[array-multiclass-classification] - assert np.float64(0.56789) == 0
FAILED tests/python_package_test/test_dask.py::test_init_score[array-regression] - assert np.float64(0.56789) == 0
FAILED tests/python_package_test/test_dask.py::test_init_score[array-ranking] - assert np.float64(0.56789) == 0
FAILED tests/python_package_test/test_dask.py::test_init_score[scipy_csr_matrix-binary-classification] - assert np.float64(0.56789) == 0
FAILED tests/python_package_test/test_dask.py::test_init_score[scipy_csr_matrix-multiclass-classification] - assert np.float64(0.56789) == 0
...
```

Reply:
Based on all of this... I feel pretty good about these changes! I think this can be merged once the other suggestions I've left have been addressed.

One other note relevant to this thread... I just pushed a new unit test to this PR: https://github.com/microsoft/LightGBM/pull/6569/commits/e76d5bc62f1d9010f4322193b3f8be567b75fa2a

That tests that the predictions from a model of all stumps takes on the correct value based on whether or not `init_score` is provided.

Right now if you train for 5 iterations but LightGBM cannot find any splits, you won't get 5 trees... you'll get 1 stump.

```python
import lightgbm as lgb
from sklearn.datasets import make_regression

X, y = make_regression(n_features=2)
bst = lgb.train(
    train_set=lgb.Dataset(X, label=y),
    params={
        "objective": "regression",
        "min_data_in_leaf": X.shape[0]
    },
    num_boost_round=5
)
bst.dump_model()["tree_info"]
```

```text
[{'tree_index': 0, 'num_leaves': 1, 'num_cat': 0, 'shrinkage': 1, 'tree_structure': {'leaf_value': 3.7640261858701707, 'leaf_count': 100}}]
```

That unit test will protect us against predictions being wrong if that ever changes and if, for example, that situation produces 5 stumps. I'm thinking about this in-progress PR, for example: #5699.

cc @jmoralez @StrikerRUS (just for awareness)

Reply:
@neNasko1 I've copied your comment from https://github.com/microsoft/LightGBM/pull/6569#issuecomment-2400950346 here ... let's please keep this in a thread, so there's only one place to link to and so it takes up less visual space scrolling through the already-very-long activity history for this PR.

---

I said:

> *When there is a init_score then boost_from_average does [not occur](https://github.com/microsoft/LightGBM/blob/06432300c0c01268c8a80c3537eef81dd5ede30d/src/boosting/gbdt.cpp#L322).*

And you asked:

> _Is this related to boost_from_average?_
> _. * When there is a init_score then boost_from_average does [not occur](https://github.com/microsoft/LightGBM/blob/06432300c0c01268c8a80c3537eef81dd5ede30d/src/boosting/gbdt.cpp#L322)._
>_. * After training the first trees(iteration=1) we then [add](https://github.com/microsoft/LightGBM/blob/06432300c0c01268c8a80c3537eef81dd5ede30d/src/boosting/gbdt.cpp#L417) the average to them._
> _...however after this fix there is a need for a change in this test._

Yes, `boost_from_average` is a part of this. That parameter is set to `true` by default

https://github.com/microsoft/LightGBM/blob/06432300c0c01268c8a80c3537eef81dd5ede30d/include/LightGBM/config.h#L946-947

And yes it's correct that when you provide an `init_score`, then boosting from the average doesn't occur, and the first tree's root node has an internal value of `0.0`.

Which test do you mean "this test" and what specific changes do you think are needed?

Reply:
Ohhhh I see. Yes I think you're right, that's a great point. Here's my proposal for making that test stricter:

1. add the following to `params` in `test_dask.py::test_init_scores` to eliminate other sources of non-determinism, so we can be more sure the difference is only due to init_score:
   - `{"seed": 708, "num_thread": 1, "deterministic": True, "force_row_wise": True
2. add a Dask equivalent of the new `test_predict_stump()` test I'd pushed: https://github.com/neNasko1/LightGBM/blob/04886c0d6d2da2e3d145916556ef34d012b43ec9/tests/python_package_test/test_engine.py#L3815-L3840

What do you think about that? And please let me know if you're not very familiar with Dask and want me to push a test like that.

Reply:
I am not that familiar with dask, but is [this](https://github.com/microsoft/LightGBM/pull/6569/commits/15fc3bfee90846a626560f4fc2acea991ca38f27) what you have in mind? 

P.S. I am running the dask tests on my M3 Mac machine by removing these [lines](https://github.com/microsoft/LightGBM/blob/master/tests/python_package_test/test_dask.py#L20-L23). Maybe it will be possible to enable `test_dask` for apple silicone in a future PR?

Reply:
I left some small suggestions for simplifying that new dask test you added, but overall ues that's exactly what I was looking for. Thanks!

> P.S. I am running the dask tests on my M3 Mac machine by removing these [lines](https://github.com/microsoft/LightGBM/blob/master/tests/python_package_test/test_dask.py#L20-L23). Maybe it will be possible to enable test_dask for apple silicone in a future PR?

Oh great to hear that the Dask tests are working for you on macOS! Thanks for letting us know. If you look in the git blame ([link](https://github.com/microsoft/LightGBM/blame/7eae66a7bd0cf5e4bb081f23b61ad11b4f66d706/tests/python_package_test/test_dask.py#L20-L23)), you'll see those markers have been there for 3 years.

<img width="1219" alt="Screenshot 2024-10-10 at 10 41 26 PM" src="https://github.com/user-attachments/assets/bf4bcbc4-9c02-4919-be35-84eea7d7b7a0">

I'll put up a separate PR trying to re-enable the Dask tests on macOS, would be great if we could re-enable them!